## Building a model lab

### Introduction

In this lesson, we'll use some data for a restaurant to see how the daily temperature can predict how many customers it will have in the day.  It records the weather each bi-weekly over two months.  

### Start with a chart of data

| temperature        | actual customers           
| ------------- |:-------------:| 
|    80       | 120 | 
|    65        | 100 | 
|    50        | 85 | 
|    70        | 100 | 
|    60        | 90 | 


Here is a list of the temperatures and a separate list of the customers.

In [1]:
temperatures = [80, 65, 50, 70, 60]

In [2]:
amounts = [120, 100, 85, 100, 90]

And we decide to use a linear regression model with the following parameters.

Our coefficient equals 1.3 and our y-intercept is 20.  Now write out the machine learning model in the cell below.  We'll get you started.

$$ number\_of\_customers = 1.3 * temperatures + 20 $$

Now use the above formula to make a prediction of our first two days, with temperatures of 80 degrees and 65 degrees.

In [3]:
expected_customers_day_1 = 1.3 * 80 + 20
expected_customers_day_1
# 124

124.0

In [4]:
expected_customers_day_2 = 1.3 * 65 + 20

expected_customers_day_2
# 104

104.5

| temperature        | actual customers | expected|
| ------------- |:-------------:| :-------------:| 
|    80       | 120 |  ? | 
|    65        | 100 |  ? | 
|    50        | 85 | 85 | 
|    70        | 100 | 111  | 
|    60        | 90 | 98 | 


### Other expressions of a model 

Now translate our machine learning model into a function in Python called `predicted_customers`.  It should be in the general form of y = mx + b.

In [5]:
def predicted_customers(x):
    return 1.3 * x + 20

In [6]:
predicted_customers(50)

# predicted_customers(50) # 85
# predicted_customers(70) # 111

85.0

In [7]:
predicted_customers(70)

111.0

### Visualizing our model

Now it's time to visualize our model.  We can start by plotting a graph of our actual data.  Remember that we alread have lists of our temperatures, and the amount of customers.

In [8]:
temperatures

[80, 65, 50, 70, 60]

In [9]:
amounts

[120, 100, 85, 100, 90]

So we can plot them, by passing them into our `trace_values` function.

In [10]:
from graph import trace_values, plot
data_trace = trace_values(temperatures, amounts)
plot([data_trace])

Now your task is display our model as trace of a line our above trace of the data.  Here is the code that we just used.  We'll guide you through in the lines that follow.

In [11]:
from graph import trace_values, plot
data_trace = trace_values(temperatures, amounts)
plot([data_trace])

So to plot a trace of our model, we need a set of x values, which are our inputs of the temperatures.

In [12]:
inputs = [50, 55, 60, 65, 70, 75, 80, 85, 90]

And it's your task to calculate the expected outputs for each of these inputs.  Assign these expected outputs to the variable `predicted_customers`.

In [13]:
type(inputs[0])

int

In [14]:
predicted_customers = [1.3 * temp + 20 for temp in inputs]
predicted_customers

[85.0, 91.5, 98.0, 104.5, 111.0, 117.5, 124.0, 130.5, 137.0]

Next we can create a `model_trace` and plot it alongside our data.

In [15]:
model_trace = trace_values(inputs, predicted_customers, mode = 'lines', name = 'model')

Uncomment the code below to display a trace of our model and our outputs.

In [16]:
from graph import trace_values, plot
data_trace = trace_values(temperatures, amounts, name = 'data')
layout = {'yaxis': {'range': [0, 140], 'title': 'expected customers'},
          'xaxis': {'title': 'temperatures'}}
plot([data_trace, model_trace], layout)

- I corrected the code "data_trace = trace_values([.30, .50, .70], [8, 11, 17], name = 'data') into as shown above cell, because the data_trace was already defined previously, as data_trace = trace_values(temperatures, amounts) to plot a graph of actual data. 
- I changed the yaxis range of [0, 18] to [0, 140] as above, because the predicted_customers has the highest value of 137 when the temperature is 90.  

### Summary

In this lesson, we built our own machine learning model and plotted it alongside our data.  We started by expressing our machine learning model as a math formula, and used it to plug in different temperatures and calculate the expected number of customers. 

Then we saw that we could translate our machine learning model into code by using a Python function that takes an input and returns an output.  Finally, we plotted our model as a line alongside our actual data.  